packages/langium/src/lsp/semantic-token-provider.ts

```diff
@@ -0,0 +1,188 @@
+            }
+        }
+        this.tokenBuilder.push(line, char, length, intType, totalModifier);
+    };
+    protected range?: Range;
+
+    semanticHighlight(document: LangiumDocument, _params: SemanticTokensParams, cancelToken: CancellationToken): SemanticTokens {
+        this.range = undefined;
+        this.resetTokensBuilder();
+        this.computeHighlighting(document, cancelToken);
+        return this.tokenBuilder.build();
+    }
+
+    semanticHighlightRange(document: LangiumDocument, params: SemanticTokensRangeParams, cancelToken: CancellationToken): SemanticTokens {
+        this.range = params.range;
+        this.resetTokensBuilder();
+        this.computeHighlighting(document, cancelToken);
+        return this.tokenBuilder.build();
+    }
+
+    semanticHighlightDelta(document: LangiumDocument, params: SemanticTokensDeltaParams, cancelToken: CancellationToken): SemanticTokens | SemanticTokensDelta {
+        this.range = undefined;
+        this.tokenBuilder.previousResult(params.previousResultId);
+        this.computeHighlighting(document, cancelToken);
+        return this.tokenBuilder.buildEdits();
+    }
+
+    protected resetTokensBuilder(): void {
+        this.tokenBuilder.previousResult(Date.now().toString());
+    }
+
+    protected computeHighlighting(document: LangiumDocument, cancelToken: CancellationToken): void {
+        const root = document.parseResult.value;
+        const treeIterator = streamAllContents(root).iterator();
+        let result: IteratorResult<AstNodeContent>;
+        do {
+            result = treeIterator.next();
+            if (!result.done) {
+                if (cancelToken.isCancellationRequested) {
+                    break;
+                }
+                const node = result.value.node;
+                const nodeRange = node.$cstNode!.range;
+                const comparedRange = this.compareRange(nodeRange);
+                if (comparedRange === 1) {
+                    break; // Every following element will not be in range, so end the loop
+                } else if (comparedRange === -1) {
+                    continue; // Current element is ending before range starts, skip to next element
+                }
+                if (this.highlightElement(node)) {
+                    treeIterator.prune();
+                }
+            }
+        } while (!result.done);
+    }
+
+    protected compareRange(range: Range | number): number {
+        if (!this.range) {
+            return 0;
+        }
+        const startLine = typeof range === 'number' ? range : range.start.line;
+        const endLine = typeof range === 'number' ? range : range.end.line;
+        if (endLine < this.range.start.line) {
+            return -1;
+        } else if (startLine > this.range.end.line) {
+            return 1;
+        } else {
+            return 0;
+        }
+    }
+
+    /**
+	 * @return true to skip the children of this element, false otherwise
```

Comment:
A boolean return type is hard to understand in this case when looking at a subclass implementation. I'd propose to either return something more specific (`void | 'prune'`) or pass the iterator as a parameter so `prune` can be called explicitly.